src/Timing.cc

```diff
@@ -0,0 +1,334 @@
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include "ord/Timing.h"
+
+#include <tcl.h>
+
+#include "db_sta/dbNetwork.hh"
+#include "db_sta/dbSta.hh"
+#include "odb/db.h"
+#include "ord/OpenRoad.hh"
+#include "sta/Search.hh"
+//#include "ord/Tech.h"
+#include "ord/Design.h"
+#include "sta/Corner.hh"
+#include "sta/Liberty.hh"
+#include "sta/TimingArc.hh"
+#include "sta/TimingRole.hh"
+#include "utl/Logger.h"
+
+namespace ord {
+
+Timing::Timing(Design* design) : design_(design)
+{
+}
+
+sta::dbSta* Timing::getSta()
+{
+  auto app = OpenRoad::openRoad();
+  return app->getSta();
+}
+
+std::pair<odb::dbITerm*, odb::dbBTerm*> Timing::staToDBPin(const sta::Pin* pin)
+{
+  ord::OpenRoad* openroad = ord::OpenRoad::openRoad();
+  sta::dbNetwork* db_network = openroad->getDbNetwork();
+  odb::dbITerm* iterm;
+  odb::dbBTerm* bterm;
+  db_network->staToDb(pin, iterm, bterm);
+  return std::make_pair(iterm, bterm);
+}
+
+bool Timing::isEndpoint(odb::dbITerm* db_pin)
+{
+  sta::Pin* sta_pin = getSta()->getDbNetwork()->dbToSta(db_pin);
+  return isEndpoint_(sta_pin);
+}
+
+bool Timing::isEndpoint(odb::dbBTerm* db_pin)
+{
+  sta::Pin* sta_pin = getSta()->getDbNetwork()->dbToSta(db_pin);
+  return isEndpoint_(sta_pin);
+}
+
+bool Timing::isEndpoint_(sta::Pin* sta_pin)
+{
+  auto search = sta::Sta::sta()->search();
+  auto vertex_array = vertices(sta_pin);
+  for (auto vertex : vertex_array) {
+    if (vertex != nullptr && search->isEndpoint(vertex)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool Timing::isStartpoint(odb::dbITerm* db_pin)
+{
+  sta::Pin* sta_pin = getSta()->getDbNetwork()->dbToSta(db_pin);
+  return isStartpoint_(sta_pin);
+}
+
+bool Timing::isStartpoint(odb::dbBTerm* db_pin)
+{
+  sta::Pin* sta_pin = getSta()->getDbNetwork()->dbToSta(db_pin);
+  return isStartpoint_(sta_pin);
+}
+
+bool Timing::isStartpoint_(sta::Pin* sta_pin)
+{
+  // TODO, Incomplete without equivalent of isEndpoint for starpoint
+  // auto search = sta::Sta::sta()->search();
+  // auto vertex_array = vertices(sta_pin);
+  // for (auto vertex : vertex_array) {
+  //  //    if (vertex != nullptr && search->isStartpoint(vertex) {
+  //  //      return true;
+  //  //    }
+  //}
+  return false;
+}
+
+float Timing::slew_corner(sta::Vertex* vertex)
+{
+  sta::Sta* sta = sta::Sta::sta();
+  float slew_max = -sta::INF;
+  for (auto corner : getCorners()) {
+    slew_max = std::max(
+        slew_max,
+        sta::delayAsFloat(sta->vertexSlew(
+            vertex, sta::RiseFall::rise(), corner, sta::MinMax::max())));
```

Comment:
do you always want max() ?

Reply:
I've added an option to pick between min and max for the slew and arrival times.